New movement system for leashed entities

peterbell10 opened the pull request:
Fixes #4146

Instead of pathfinding towards the leashed to entity, monsters are now accelerated towards the leashed to entity as if they were attached by a spring.

Here's a gif of it in action: https://gfycat.com/DaringUniformAtlasmoth

bearbin commented on src/Mobs/Monster.cpp:
```diff
@@ -435,19 +428,41 @@ void cMonster::CalcLeashActions()
+		return;
+	}
+
+	static const double LeashNaturalLength = 5.0;   // Length at equilibrium
```
The comment here suggests that when entities are closer than the NaturalLength, they are pushed away. Perhaps "Length where the leash begins to pull" would be better? That doesn't sound great either so idk.

peterbell10 replied:
Yes, equilibrium isn't quite right here.  I'll change it.

bearbin commented on src/Mobs/Monster.cpp:
```diff
@@ -435,19 +428,41 @@ void cMonster::CalcLeashActions()
+	{
+		// Accelerate monster towards the leashed to entity:
+		auto Extension = Distance - LeashNaturalLength;
+		auto Direction = Displacement / Distance;
```
Is this not just `Displacement.Normalize()`?

peterbell10 replied:
Yes direction is the normalised displacement, I was just factoring out the length calculation as it's needed elsewhere.  I can change it to `NormalizeCopy` if you prefer.

bearbin commented on src/Mobs/Monster.cpp:
```diff
@@ -435,19 +428,41 @@ void cMonster::CalcLeashActions()
 		}
+
+		// Apply the acceleration
+		using namespace std::chrono;
```
Does this really add enough to justify the extra line?

peterbell10 replied:
I find it more readable this way.  Cuberite's style conventions don't tend to be economic with line counts anyway.

bearbin commented on src/Mobs/Monster.cpp:
```diff
@@ -435,19 +428,41 @@ void cMonster::CalcLeashActions()
+
+	static const double LeashNaturalLength = 5.0;   // Length at equilibrium
+	static const double LeashMaximumLength = 10.0;  // Length where the leash breaks
+	static const double LeashSpringConstant = 20.0;  // How elastic the leash is
```
Similarly here, a higher value means less elastic, which is not the intuiitive meaning of the comment.

peterbell10 replied:
I'm struggling to come think of language that isn't technical. It makes sense to me because the spring constant is proportional to the [elastic modulus](https://en.wikipedia.org/wiki/Young%27s_modulus).

Would "stiff" make more sense to you?

peterbell10 commented:
I took another look at the behaviour and realised it was a bit strange when the mob doesn't try to pathfind towards the player and just gets dragged behind. The reason I removed it in the first place was I didn't like how the mobs would stand on top of the player.  So instead of removing the pathing, I've fixed it with the `CloseFollowDistance`.

bearbin commented on src/Mobs/Monster.cpp:
```diff
@@ -435,19 +412,54 @@ void cMonster::CalcLeashActions()
+	{
+		const Vector3d TargetBlock((LeashedToPos - Direction * CloseFollowDistance).Floor());
+		// Move to centre of target block face
+		MoveToPosition(TargetBlock + Vector3d{ 0.5, 0.0, 0.5 });
```
I'm not too familiar with the entity movement calculations, but is this movement cancelled when the leashed entity gets within the CloseFollowDistance? I feel like this could end up with mobs sitting on top of the player again.

peterbell10 replied:
You're right that the movement isn't cancelled because `MoveToPosition` isn't called.  However the mobs aren't pathing towards the block the player is standing on in the first place. Before rounding, the target block is the nearest point on a sphere of radius `CloseFollowDistance` centred on the the leashed to entity.

If you're wondering why the if is necessary then: it actually stops leashed mobs from walking away when the player walks within the `CloseFollowDistance`

bearbin replied:
Thanks for explaining, I understand it now and the logic looks good.

bearbin commented:
The OSX builders on travis appear to be broken - but the build has passed otherwise for this PR.

peterbell10 commented:
https://www.traviscistatus.com/ claims the build workers were broken but should be fixed and working through a backlog.

I would merge this but I don't have permission to merge without CI passing.  If you want to merge @bearbin, go ahead.
